File: Classes/ThemeManager/LWThemeManager.m
//
//  LWThemeManager.m
//  LWTheme
//
//  Created by lalawue on 2019/12/8.
//  Copyright © 2019 lalawue. All rights reserved.
//

#import "LWThemeManager_Private.h"
#import "UIColor+HexString.h"

NSString *const kLWThemeModeUpdateNotification = @"kLWThemeModeUpdateNotification";

@interface LWThemeManager ()
@property (nonatomic, strong) NSDictionary *colorMap;
@property (nonatomic, strong) NSDictionary *fontMap;
@property (nonatomic, strong) NSDictionary *imageMap;
@property (nonatomic, strong) NSHashTable *themeObjects; ///< object applly theme color
@end

@implementation LWThemeManager

+ (instancetype)sharedInstance
{
    static LWThemeManager *manager;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        manager = [[LWThemeManager alloc] init];
        manager.themeObjects = [NSHashTable weakObjectsHashTable];
    });
    return manager;
}

- (BOOL)isDarkMode
{
    return self.themeMode == LWThemeMode_Dark;
}

- (void)useThemeMode:(LWThemeMode)themeMode withThemeMap:(NSDictionary *)themeMap
{
    if ([themeMap isKindOfClass:[NSDictionary class]]) {
        _themeMode = themeMode;
        self.colorMap = themeMap[@"Color"];
        self.fontMap = themeMap[@"Font"];
        self.imageMap = themeMap[@"Image"];
        [self changeThemeForObjects];
        [[NSNotificationCenter defaultCenter] postNotificationName:kLWThemeModeUpdateNotification object:@(themeMode)];
    }
}

- (void)changeThemeForObjects
{
    if (self.themeObjects.count) {
        NSArray *objects = [self.themeObjects.allObjects copy];
        for (id<LWThemeObjectNotify> obj in objects) {
            if ([obj respondsToSelector:@selector(lw_callSetterWithValue)]) {
                [obj lw_callSetterWithValue];
            }
        }
    }
}

- (UIColor *)colorForPattern:(NSString *)colorPattern
{
    if (colorPattern.length) {
        NSString *colorString = [self.colorMap objectForKey:colorPattern];
        UIColor *colorValue = [UIColor colorWithHexString:colorString.length ? colorString : colorPattern];
        return colorValue;
    }
    return nil;
}

- (NSString *)colorStringForPattern:(NSString *)colorPattern
{
    return self.colorMap[colorPattern];
}

- (UIFont *)fontForPattern:(NSString *)fontPattern
{
    NSString *fontDesc = self.fontMap[fontPattern];
    if (fontDesc.length) {
        NSRange r = [fontDesc rangeOfString:@":"];
        if (r.location != NSNotFound) {
            NSString *fontFamily = [fontDesc substringToIndex:r.location];
            CGFloat fontSize = [[fontDesc substringFromIndex:r.location + 1] floatValue];
            if (fontFamily.length<=0 || [fontFamily isEqualToString:@"_"]) {
                return [UIFont systemFontOfSize:fontSize];
            }
            if ([fontFamily.lowercaseString isEqualToString:@"bold"]) {
                return [UIFont boldSystemFontOfSize:fontSize];
            }
            if ([fontFamily.lowercaseString isEqualToString:@"italic"]) {
                return [UIFont italicSystemFontOfSize:fontSize];
            }
            return [UIFont fontWithName:fontFamily size:fontSize] ?: [UIFont systemFontOfSize:fontSize];
        }
    }
    return [UIFont systemFontOfSize:12];
}

- (UIImage *)imageNamed:(NSString *)imagePattern
{
    if (imagePattern.length) {
        NSString *imageName = self.imageMap[imagePattern];
        if (imageName) {
            return [UIImage imageNamed:imageName];
        } else {
            return [UIImage imageNamed:imagePattern];
        }
    }
    return nil;
}

@end

#pragma mark -

@implementation LWThemeManager (Notifier)

- (void)addThemeObject:(id<LWThemeObjectNotify>)object
{
    if ([object respondsToSelector:@selector(lw_callSetterWithValue)]) {
        [self.themeObjects addObject:object];
    }
}

- (void)removeThemeObject:(id)object
{
    [self.themeObjects removeObject:object];
}

@end

File: Classes/ThemeManager/LWThemeManager.swift
//
//  LWThemeManager.swift
//  LWThemeSwift
//
//  Created by lalawue on 2019/12/17.
//  Copyright © 2019 lalawue. All rights reserved.
//

import UIKit

public let kLWThemeModeUpdateNotification = "kLWThemeModeUpdateNotification" ///< notification name

public enum LWThemeMode {
    case Light
    case Dark
}

public class LWThemeManager {

    public static let sharedInstance = LWThemeManager();
    
    public var themeMode : LWThemeMode {
        return _themeMode;
    }
    
    public var isDarkMode : Bool {
        return _themeMode == LWThemeMode.Dark
    }
    
    public func useThemeMode(_ mode : LWThemeMode, themeMap : NSDictionary) {
        _themeMode = mode
        _colorMap = themeMap["Color"] as? [String : String] ?? [:]
        _fontMap = themeMap["Font"] as? [String : String] ?? [:]
        _imageMap = themeMap["Image"] as? [String : String] ?? [:]
        changeThemeForObjects()
        NotificationCenter.default.post(name: NSNotification.Name(rawValue: kLWThemeModeUpdateNotification), object: mode)
    }
    
    public func colorForPattern(_ colorPattern : String?) -> UIColor? {
        if (colorPattern == nil) {
            return nil
        }
        let colorString = _colorMap[colorPattern!] ?? colorPattern!
        let color = UIColor.colorFromHexString(hexString: colorString)
        return color
    }
    
    public func colorStringForPattern(_ colorPattern : String) -> String? {
        return _colorMap[colorPattern]
    }
    
    public func fontForPattern(_ fontPattern : String?) -> UIFont? {
        if (fontPattern != nil) {
            let fontDesc = _fontMap[fontPattern!]
            if (fontDesc != nil && fontDesc!.count > 0) {
                let array = fontDesc!.split(separator: ":")
                if (array.count >= 2) {
                    let fontName = array[0]
                    let fontSize = CGFloat.init( Double(array[1])! )
                    if (fontName.count <= 0 || fontName.elementsEqual("_")) {
                        return UIFont.systemFont(ofSize: fontSize)
                    }
                    if (fontName.lowercased().elementsEqual("bold")) {
                        return UIFont.boldSystemFont(ofSize: fontSize)
                    }
                    if (fontName.lowercased().elementsEqual("italic")) {
                        return UIFont.italicSystemFont(ofSize: fontSize)
                    }
                    return UIFont.init(name: String(fontName), size: fontSize)!
                }
            }
            return UIFont.systemFont(ofSize: 12)
        } else {
            return nil
        }
    }
    
    public func imageNamed(_ name : String? ) -> UIImage? {
        if (name != nil) {
            let imageName = _imageMap[name!]
            if (imageName != nil) {
                return UIImage.init(named: imageName!)
            } else {
                return UIImage.init(named: name!)
            }
        }
        return nil
    }
    
    //MARK:
    
    func addThemeObject(_ themeObject : LWThemeObjectNotify) {
        if (themeObject.responds(to: #selector(LWThemeObjectNotify.lw_callSetterWithValue))) {
            _themeObjects.add(themeObject)
        }
    }
    
    func removeThemeObject(_ themeObject : LWThemeObjectNotify ) {
        _themeObjects.remove(themeObject)
    }
    
    //MARK:

    private init() {
        _themeMode = LWThemeMode.Light;
    }

    fileprivate var _themeMode : LWThemeMode;
    fileprivate var _colorMap = [String : String]()
    fileprivate var _fontMap = [String : String]()
    fileprivate var _imageMap = [String : String]()
    fileprivate var _themeObjects = NSHashTable<AnyObject>(options: NSHashTableWeakMemory)
    
    func changeThemeForObjects() {
        if (_themeObjects.count <= 0) {
            return
        }
        let allObjects = _themeObjects.allObjects
        for obj in allObjects {
            let themeObject = obj
            if (themeObject.responds(to: #selector(LWThemeObjectNotify.lw_callSetterWithValue))) {
                themeObject.lw_callSetterWithValue()
            }
        }
    }
}
